### ISSUE TITLE
Resolve entry-point assembly on profiler initialization

### ISSUE THREAD

**Issue by iskiselev** [dotnet/coreclr#15669]
Looks like there is no any simple way to resolve entry-point assembly path in `ICorProfilerCallback::Initialize` to take a decision, if application should be profiled or not in case if application is started using dotnet. The only way will I see now is to parse command-line arguments within profiler, which is bad - as some command-line arguments treated as "internal implementation details" (see dotnet/cli#2243) and in any case require repeating many part of CoreCLI command-line parser.
I'm not sure which repository is best for this issue. It looks related to dotnet/core-setup#2007 and some discussions in dotnet/cli#6237.

**Comment by jkotas** [comment]
cc @sywhang @noahfalk

**Comment by noahfalk** [comment]
Yes, as-is the API doesn't provide that info, though other profilers have never been structured in a way that they needed it. I think a typical workflow is that the profiler vendor creates MyProfilerWithUI.exe which a user launches. The user indicates they want to profile foo.exe, and then MyProfilerWithUI.exe launches "dotnet.exe foo.exe" on their behalf with the environment variables set to enable profiling. The runtime observes the environment variable is set, loads the profiling dll, and in initialize there is no further decision making that needs to occur. There might be other implementation options depending on what scenario you are trying to achieve.

In terms of changing the product to give you the entrypoint assembly in profiler initialize I don't see any way to do it without either specializing it for specific hosts or changing the hosting API. If you use coreconsole.cpp as an example, profiler initialize occurs inside of host->Start()
https://github.com/dotnet/coreclr/blob/master/src/coreclr/hosts/coreconsole/coreconsole.cpp#L429
but the entrypoint assembly isn't provided until later in host->ExecuteAssembly()
https://github.com/dotnet/coreclr/blob/master/src/coreclr/hosts/coreconsole/coreconsole.cpp#L529
The runtime can not give the profiler information that it has not yet received from the host.

A little more technically palatable might be APIs that provide the profiler dll the identity of the entrypoint assembly once it has been provided to the runtime, but without knowing more about what you are trying to accomplish I don't know if it meets your needs. Certainly there are some profiling actions you can only do at startup, but others can be enabled/disabled after the fact. Simply loading a profiler dll into a running .Net app doesn't on its own make significant changes in how the application will run. A user might still perceive the app to be 'not profiled' even if the profiler dll did load/initialize was called.

I'm going to change this to milestone future as this is just an open discussion so far.
